Fixing cooling mode cooling coil cooling rate much higher than heat pump cooling rate and OU capacity

yujiex opened the pull request:
Pull request overview
---------------------
 - Fixes #10187 
 - Fixes #10607

For both the original and the fluid-control VRF model, terminal unit capacity should have been constrained by condenser capacity, where a maximum allowable terminal unit capacity is calculated in `LimitTUCapacity(.)` for each VRF system. The maximum allowable terminal unit capacity is stored in `state.dataHVACVarRefFlow->MaxCoolingCapacity(VRFCond)` and passed into `SimDXCoil(.)` which is NOT passed into `CalcVRFCoolingCoil_FluidTCtrl(.)`, as compared with the non-fluid-control VRF model, `CalcVRFCoolingCoil(.)`

This PR will add an optional argument `MaxCoolCap` to `CalcVRFCoolingCoil_FluidTCtrl(.)` as is in CalcVRFCoolingCoil(.) and use this quantity to constrain the cooling coil capacity.
 
The following shows the effect of the fix: before the fix, cooling coil cooling rate is much higher than heat pump cooling rate; after the fix, the two cooling rate are similar. The test file used here is the US+SF+CZ4A+hp+crawlspace+IECC_2006_VRF.idf example file, using NY weather epw input. The "Rated Evaporative Capacity" of the "AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl" is adjusted to 500W.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/fe869c25-d3d6-42f3-9d9b-49f2bd4bbac9)

### Regression diff

The following files have regression diff
- DOASDXCOIL_wADPBFMethod
- DOASDXCOIL_wADPBFMethod_NoReturnPath
- DOAToVRF
- HVACTemplate-5ZoneVRF
- US+SF+CZ4A+hp+crawlspace+IECC_2006_VRF
- VRFMultispeedFan
- VariableRefrigerantFlow_5Zone
- VariableRefrigerantFlow_5Zone_wAirloop
- VariableRefrigerantFlow_FluidTCtrl_5Zone
- VariableRefrigerantFlow_FluidTCtrl_HR_5Zone
- VariableRefrigerantFlow_FluidTCtrl_wSuppHeater_5Zone
- VariableRefrigerantFlow_wSuppHeater_5Zone
- _DOASDXCOIL_wUserSHRMethod

#### ESO difference

The following files have ESO diffs in Zone VRF Air Terminal Total(Latent) Cooling(Heating) Rate. These are expected as a result of the changes in the total and latent system output calculation which fixed the problem of total != sensible + latent output.

- DOASDXCOIL_wADPBFMethod
- DOASDXCOIL_wADPBFMethod_NoReturnPath
- DOAToVRF
- HVACTemplate-5ZoneVRF
- VRFMultispeedFan
- VariableRefrigerantFlow_5Zone
- VariableRefrigerantFlow_5Zone_wAirloop
- VariableRefrigerantFlow_wSuppHeater_5Zone
- _DOASDXCOIL_wUserSHRMethod

The 4 files with fluidTCtrl VRF model (
US+SF+CZ4A+hp+crawlspace+IECC_2006_VRF, VariableRefrigerantFlow_FluidTCtrl_5Zone, VariableRefrigerantFlow_FluidTCtrl_HR_5Zone, VariableRefrigerantFlow_FluidTCtrl_wSuppHeater_5Zone
) have difference in many more fields as follows. The core difference is the coil cooling rate and the VRF air terminal cooling rate. The cooling coil cooling rate difference (due to a capacity limit added in this feature) will also cause difference in compressor side cooling output, such as cooling rate, cooling electricity rate, etc. Piping correction is also affected as it's calculated as (coil cooling rate) / (coil cooling rate + piping loss). The coil cooling rate difference will also cause the temperature and air flow difference in corresponding nodes. The coil cooling rate difference will affect zone temperature as well (In the extreme case, when OU capacity is very small, the zone might not be able to control to setpoint temperature any more. This won't happen if coil has unlimited capacity like before the fix). Zone temperature difference would result in slight difference in heating coil and supplemental heating coil heating rate as well.

  - Zone Air Temperature 
- Zone Air Relative Humidity 
- Zone Air Temperature 
- Zone Air Relative Humidity 
- Zone Air System Sensible Heating Rate 
- Zone Air System Sensible Cooling Rate 
- Zone Predicted Sensible Load to Setpoint Heat Transfer Rate 
- Zone Predicted Sensible Load to Heating Setpoint Heat Transfer Rate 
- Zone Predicted Sensible Load to Cooling Setpoint Heat Transfer Rate 
- Fan Electricity Rate 
- Cooling Coil Total Cooling Rate 
- Cooling Coil Sensible Cooling Rate 
- Cooling Coil Latent Cooling Rate 
- Cooling Coil Runtime Fraction 
- Zone VRF Air Terminal Cooling Electricity Rate 
- Zone VRF Air Terminal Cooling Electricity Energy 
- Zone VRF Air Terminal Total Cooling Rate 
- Zone VRF Air Terminal Sensible Cooling Rate 
- Zone VRF Air Terminal Latent Cooling Rate 
- Zone VRF Air Terminal Heating Electricity Rate 
- Zone VRF Air Terminal Heating Electricity Energy 
- Zone VRF Air Terminal Total Heating Rate 
- Zone VRF Air Terminal Sensible Heating Rate 
- Zone VRF Air Terminal Latent Heating Rate 
- VRF Heat Pump Total Cooling Rate 
- VRF Heat Pump Total Heating Rate 
- VRF Heat Pump Cooling Electricity Rate 
- VRF Heat Pump Cooling Electricity Energy 
- VRF Heat Pump Heating Electricity Rate 
- VRF Heat Pump Heating Electricity Energy 
- VRF Heat Pump Compressor Electricity Rate 
- VRF Heat Pump Compressor Rotating Speed 
- VRF Heat Pump Outdoor Unit Condensing Temperature 
- VRF Heat Pump Outdoor Unit Evaporating Temperature 
- System Node Standard Density Volume Flow Rate 
- System Node Wetbulb Temperature 
- System Node Standard Density Volume Flow Rate 

#### EIO difference

EIO difference are in the four idfs with FluidTCtrl model. There are some warmup differences in each zone in the model. This is expected as the coil cooling rate difference will change the zone temperature, which affects the warmup convergence.

#### Meter diff

Meter diffs happen in two of the idfs with VRF FluidTCtrl object. It is expected as the cooling coil cooling rate difference will also affect its electricity rate, and corresponding HVAC, Facility electricity rate part of which is coil electricity rate.

#### Table string diffs

These table diffs happen in the four idfs with VRF FluidTCtrl objects. The diffs are in warmup statistics and electricity rate, which are explained in the eio diff and meter diff sections.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
initial commit of example file

yujiex committed:
fix cooling with similar approach as heating

yujiex committed:
add testfile to CMakeList, add weather file, clang-format

yujiex committed:
test file for cooling not meeting demand case

Commit:
use the AirMassFlow like in heating mode, remove TU_CoolingLoad_actual

Follow the heating mode case to use the air flow rate calculated from the
ControlVRFIUCoil, this makes coil cooling rate bounded by the OU capacity,
although the heat pump cooling rate and the coil cooling rate are still not
exactly the same.

Piping correction doesn't need to bound the load either, similar to the case in
heating mode

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

Commit:
remove duplicated add testfile in CMakeLists

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -231,7 +231,7 @@ void SimDXCoil(EnergyPlusData &state,
     } break;
     case HVAC::CoilVRF_FluidTCtrl_Cooling: {
-        CalcVRFCoolingCoil_FluidTCtrl(state, DXCoilNum, HVAC::CompressorOp::On, FirstHVACIteration, PartLoadRatio, fanOp, CompCycRatio, _, _);
+        CalcVRFCoolingCoil_FluidTCtrl(state, DXCoilNum, HVAC::CompressorOp::On, FirstHVACIteration, PartLoadRatio, fanOp, CompCycRatio, _, _, MaxCap);
```
I'm not sure why the max capacity is not always passed into the coil. It's either 1.0e+20 or it's a value. That's for another day.

yujiex replied:
I just changed MaxCoolCap to regular non-optional arguments. @rraustad

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -11374,9 +11374,6 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state)
 
-            Q_c_TU_PL = CompEvaporatingCAPSpdMax;
-            TU_CoolingLoad = CompEvaporatingCAPSpdMax;
-            this->TUCoolingLoad = TU_CoolingLoad;
```
Should TU_CoolingLoad be reset here or not? This var is used at line 11385 below. For a VRF system with 1 TU this might make sense but for multiple TUs this will likely have no impact.

yujiex replied:
It shouldn't be reset here. Similar to the situation in the heating side, which we fixed a while ago: https://github.com/NREL/EnergyPlus/pull/10154
I copied in the relative discussion here:
![image](https://github.com/NREL/EnergyPlus/assets/13421161/a72e8046-fbe6-4fa4-ab10-b9cf09bf68cb)

This upper bounding will cause the compressor to reduce its output at more extreme weather, which is bad. So this upper bounding needs to be removed.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -11540,9 +11537,9 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state)
-        state.dataHVACVarRefFlow->MaxHeatingCapacity(VRFCond) = 0.0;
+        this->HeatingCapacity = 0.0;                                 // Include the piping loss
+        this->PipingCorrectionHeating = 1.0;                         // 1 means no piping loss
+        state.dataHVACVarRefFlow->MaxHeatingCapacity(VRFCond) = 0.0; // yujie: might be wrong here too, should be MaxCap = 1e+20
```
Please check where MaxHeatingCapacity and MaxCoolingCapacity get set to MaxCap(1.0e+20). That will answer your question. It should either be set to MaxCap or not set at all. Maybe this is the reason that MaxCap is an optional argument, because it's 0 at times. If this var were never set to 0, and only set to MaxCap when reset is needed, and only reduced when the system is overloaded, then MaxCap could always be passed to the coil model and NOT be an optional argument.

yujiex replied:
I just changed MaxCoolCap to non-optional.
I think this initialization here (`state.dataHVACVarRefFlow->MaxHeatingCapacity(VRFCond) = 0.0`) might be just an error. The default for capacity limit should be a really large number, so that no limit is placed.

dareumnam replied:
Have you checked where `MaxHeatingCapacity` and `MaxCoolingCapacity` are set to MaxCap(1.0e+20)? Or if you believe this initialization is an error, I think you can correct the code line and delete the comment in this code.

yujiex replied:
It should be 1.0e+20. I'll change it and remove the comment.

yujiex replied:
@dareumnam. I corrected the code line and removed the comment on the 4th. Sorry I forgot to ping you. Please see if it looks okay. Thanks

rraustad commented:
This change did improve the results based on the data shown in the description. This may be for a different issue, but how can coil capacity ever exceed HP capacity? Isn't it sum of coil capacity + piping losses = HP capacity? So coil capacity should always be less than or equal to HP capacity.

yujiex commented:
> This change did improve the results based on the data shown in the description. This may be for a different issue, but how can coil capacity ever exceed HP capacity? Isn't it sum of coil capacity + piping losses = HP capacity? So coil capacity should always be less than or equal to HP capacity.

Indeed, I'll look into this. There might still be something not quite right. Coil cooling rate should be no more than heat pump cooling rate

Commit:
Change MaxCoolCap to non-optional arg

yujiex commented:
> This change did improve the results based on the data shown in the description. This may be for a different issue, but how can coil capacity ever exceed HP capacity? Isn't it sum of coil capacity + piping losses = HP capacity? So coil capacity should always be less than or equal to HP capacity.

I investigated this and found the issue might be caused by this 0.65 min fan speed ratio constraint. 

```
    FanSpdRatioMin = min(max(OAMassFlow / Garate, 0.65), 1.0); // ensure that coil flow rate is higher than OA flow rate
```

After removing this constraint, the difference become smaller and the coil cooling rate is slightly smaller than HP cooling rate

![image](https://github.com/NREL/EnergyPlus/assets/13421161/5741fc3a-eb25-4ffd-bef2-76359d7b141d)

@rraustad

Commit:
make MaxCap and MaxCoolCap a default argument, move constant declare

move the MaxCap(1e+20) to DataGlobalConstants, to avoid circular inclusion issue

Commit:
remove unwanted comment

Myoldmopar commented:
@yujiex in addition to the diffs in a handful of files, there are 3 unit tests that are now failing and must be dealt with before this goes much further.  And at least one of the test failures is quite a significant change.  We can discuss on technicalities if needed.

yujiex commented:
> @yujiex in addition to the diffs in a handful of files, there are 3 unit tests that are now failing and must be dealt with before this goes much further. And at least one of the test failures is quite a significant change. We can discuss on technicalities if needed.

Hi @Myoldmopar. The unit test failing is caused by this change (if you revert this part, the unit test can run through).

![image](https://github.com/NREL/EnergyPlus/assets/13421161/1369132f-bce4-44f6-af47-a87f97f7c481)

Previously the coil TotalCoolingEnergyRate and SensCoolingEnergyRate uses the InletAirMassFlowRate (shown in the following code chunk) instead of the AirMassFlow computed using the `FanSpdRatio` output from function `ControlVRFIUCoil`

```
        Real64 AirMassFlowRate = thisDXCoil.InletAirMassFlowRate;
        // Coil total/sensible/latent cooling rates
        CalcComponentSensibleLatentOutput(AirMassFlowRate,
                                          InletAirDryBulbTemp,
                                          InletAirHumRat,
                                          OutletAirTemp,
                                          OutletAirHumRat,
                                          thisDXCoil.SensCoolingEnergyRate,
                                          thisDXCoil.LatCoolingEnergyRate,
                                          thisDXCoil.TotalCoolingEnergyRate);

```

This might be wrong. The air flow should be changed to meet the cooling demand.

For example, in heating mode, the heating energy rate is actually computed using the `AirMassFlow` from function `ControlVRFIUCoil`

```
        ControlVRFIUCoil(state,
                         DXCoilNum,
                         QCoilReq,
                         thisDXCoil.InletAirTemp,
                         thisDXCoil.InletAirHumRat,
                         thisDXCoil.CondensingTemp,
                         AirMassFlowMin,
                         FanSpdRatio,
                         OutletAirHumRat,
                         OutletAirTemp,
                         OutletAirEnthalpy,
                         ActualSH,
                         ActualSC);
        AirMassFlow = FanSpdRatio * thisDXCoil.RatedAirMassFlowRate(Mode);
        ...
        thisDXCoil.TotalHeatingEnergyRate = AirMassFlow * (OutletAirEnthalpy - InletAirEnthalpy) * PartLoadRatio;
```

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

Commit:
need to multiply in PLR in the cooling coil energy calculation

The air mass flow need to multiply by the part load ratio like in the inlet node
air mass flow calculation

Also in heating mode, the energy calculation does have a PLR terminal

thisDXCoil.TotalHeatingEnergyRate = AirMassFlow * (OutletAirEnthalpy -
InletAirEnthalpy) * PartLoadRatio;

Myoldmopar commented:
@yujiex does this have any more status updates after the technicalities call last week?

yujiex commented:
> @yujiex does this have any more status updates after the technicalities call last week?

Hi @Myoldmopar , I am working on resolving this unit test "VRF_FluidTCtrl_ReportOutputVerificationTest". The other two has been fixed. For this unit test, I found that in addition to the air mass flow rate variable cause I have identified last time, this line below also fails (mismatching between coil and TU cooling rate). I am trying to figure out the source and try to fix it today.
```
    EXPECT_NEAR(thisDXCoolingCoil.TotalCoolingEnergyRate, (thisVRFTU.TotalCoolingRate + thisFan->totalPower), 0.0001);
```

Commit:
fix sensible and latent load calculation in VRF model

LatentLoad Met should be like in GeneralRoutines.cc like this

TotalOutput = MassFlow * (Psychrometrics::PsyHFnTdbW(TDB2, W2) - Psychrometrics::PsyHFnTdbW(TDB1, W1)); // total addition/removal rate, {W};
SensibleOutput = MassFlow * Psychrometrics::PsyDeltaHSenFnTdb2W2Tdb1W1(TDB2, W2, TDB1, W1); // sensible addition/removal rate, {W};
LatentOutput = TotalOutput - SensibleOutput;                                                // latent addition/removal rate, {W}

TotalOutput - SensibleOutput does not equal to this

(AirMassFlow * (SpecHumOut - SpecHumIn)) * PsyHgAirFnWTdb(0.0, TempOut)

using this to calculate latent output will make

TotalOutput != SensibleOutput + LatentOutput

Commit:
Keep OAMassFlow/Garate as lower bound of FanSpdRatio

Commit:
clang-format

Commit:
output correct latent load

yujiex commented:
@Myoldmopar I fixed the unit tests.

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

Commit:
Change error message to allow only Fan:SystemModel

Commit:
clang-format

Commit:
Revert "Change error message to allow only Fan:SystemModel"

This reverts commit b8c13b9ea37bc3dbac8b178e4db6bd3e82dc7c8e.
commited to the wrong branch

yujiex commented:
Hi @Myoldmopar, would you mind taking another look? Thanks.

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

nrel-bot-2c commented:
@Myoldmopar @yujiex @myoldmopar it has been 28 days since this pull request was last updated.

yujiex committed:
Merge branch 'VRFFluidFixCoolingTUCap' of https://github.com/NREL/EnergyPlus into VRFFluidFixCoolingTUCap

dareumnam committed:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

Commit:
change default MaxHeatingCapacity to 1e20, pipingCorrection default

default for piping correction should be 1 (no effect)

Commit:
Merge branch 'VRFFluidFixCoolingTUCap' of https://github.com/NREL/EnergyPlus into VRFFluidFixCoolingTUCap

Commit:
fix unit test and clang-format

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

dareumnam commented:
@rraustad, Do you have any further comments on this? I think it's good to go after @Myoldmopar's final review, unless you have additional feedback.

yujiex commented:
@Myoldmopar I will resolve the conflict in a bit (on bus now)

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

yujiex commented:
@Myoldmopar I just resolved the conflict.

Myoldmopar commented:
Thanks, let's see how CI responds in like 30 minutes.

Myoldmopar commented:
![image](https://github.com/user-attachments/assets/f60e8580-16ac-401c-b89f-2627bb00b031)

OK OK 35 minutes not 30.  

So there are 2 more files now showing big diffs.  I wonder what's caused that.  Do you have any idea?

Commit:
Merge remote-tracking branch 'origin/develop' into VRFFluidFixCoolingTUCap

yujiex commented:
@Myoldmopar It's strange, as the additional diffs happen in "CentralChillerHeaterSystem_Cooling_Heating" and "CentralChillerHeaterSystem_Simultaneous_Cooling_Heating", which doesn't have the VRF objects. Could it be because I'm not up-to-date with the develop?

Myoldmopar commented:
Hmmm.  Now that you mention it, maybe so.  I was just bragging about how we wouldn't see diffs just from being behind develop.  But maybe there's still a way.  When the CI system is doing the build, it checks out the current develop as the baseline.  Because CI is a bit bogged down right now, develop has definitely been moving fast.  So yeah, it's possible.  In most scenarios it won't be happening, just if develop moves really really fast.

yujiex commented:
> Hmmm. Now that you mention it, maybe so. I was just bragging about how we wouldn't see diffs just from being behind develop. But maybe there's still a way. When the CI system is doing the build, it checks out the current develop as the baseline. Because CI is a bit bogged down right now, develop has definitely been moving fast. So yeah, it's possible. In most scenarios it won't be happening, just if develop moves really really fast.

seems develop is getting many things merged in today

Myoldmopar commented:
Back to the 11 ESO big diffs like above.  Seems like that took care of it.  Thanks @yujiex

yujiex commented:
Thanks @Myoldmopar

Myoldmopar committed:
Merge develop; resolve conflicts

Myoldmopar commented:
I resolved the test conflicts and everything is passing locally.  I'm going to stop for lunch and let CI catch up and get fresh results to confirm I've got this ready.  If CI is clean this can go in.

yujiex commented:
Thanks for handling the conflict @Myoldmopar

Myoldmopar commented:
Looks like I didn't break anything with my conflict resolution, merging this.  Thanks @yujiex
